design/complex-workloads.md

```diff
@@ -0,0 +1,988 @@
+    provider: gcp
+  resources:
+  - name: demo
+    secretName: demo
+  targetDeployment:
+    apiVersion: extensions/v1beta1
+    kind: Deployment
+    metadata:
+      name: wordpress
+      labels:
+        app: wordpress
+    spec:
+      selector:
+        app: wordpress
+      template:
+        metadata:
+          labels:
+            app: wordpress
+        spec:
+          containers:
+            - name: wordpress
+              image: wordpress:4.6.1-apache
+              ports:
+                - containerPort: 80
+  targetNamespace: demo
+  targetService:
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: wordpress
+    spec:
+      ports:
+        - port: 80
+      selector:
+        app: wordpress
+      type: LoadBalancer
+```
+
+Workloads are modeled in Crossplane 0.1 as a Custom Resource Definition[12]
+(CRD) embedding a Kubernetes `Namespace`, `Deployment` and `Service` -
+`.spec.targetNamespace`, `.spec.targetDeployment` and `.spec.targetService`
+respectively. Once the scheduler has scheduled the `Workload` to a cluster the
+workload controller connects to said cluster and creates the templated
+`Deployment` and `Service`. The controller polls the status of the `Deployment`
+and `Service` during its sync phase, persisting them inline in the `Workload`'s
+`.status` field. Each `Workload` may also contain a set of references to
+Crossplane resources or resource claims upon which the Workload depends -
+modeled as distinct Kubernetes resources - in order to replicate their
+connection `Secrets` to the cluster upon which the `Workload` is scheduled.
+
+Complex applications such as Gitlab exceed the capabilities of today's
+`Workload` resource. Gitlab recommends deploying to Kubernetes via Helm. When
+configured to use managed services for Redis, SQL, and Buckets the chart renders
+to almost 4,800 lines of YAML including 14 `Deployments`, 1 `StatefulSet`,
+3 `Jobs`, 9 `Services`, 16 `ConfigMaps`, and many other resources. Crossplane
+must be able to model complex applications as complex workloads.
+
+
+## Goals
+The goal of this document is to design _part_ of the best possible user
+experience for deploying complex applications with Crossplane; `Workload` will
+not be responsible for the entire application installation and lifecycle
+management but rather be a building block that may be managed by higher level
+constructs.
+
+It is important that:
+* Workloads can model any Kubernetes resource, including built in resources and
+  those defined by CRDs.
+* Users do not need to connect to the cluster to which a `Workload` is scheduled
+  in order to determine the status of the resources (`Deployments`, etc) managed
+  by said `Workload`.
+* Each `Workload` is a unit of scheduling; it may not be spread across multiple
+  `KubernetesClusters`.
+* The proposed design lay a foundation for supporting workloads that are not
+  containerised.
+
+The following are out of scope for the Workload resource:
```

Comment:
Much of our internal debate has related to whether or not workloads should only target fresh clusters and/or namespaces. Put otherwise, whether installing a Workload in an existing Kubernetes cluster and/or namespace is a use case we should design for.

We should reach alignment on this and make it explicit. This might take the form of a 'guiding principle' rather than a strict goal or non-goal.